tree/10_ncf_internals/list-compatible-inputs

```diff
@@ -62,31 +93,68 @@ do
+
+      # @agent_version
       if grep -q "${tag_regex}>=" "${file}"; then operator=">="; fi
       if grep -q "${tag_regex}<" "${file}"; then operator="<"; fi
-      major=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\1/p" "${file}"`
-      minor=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\2/p" "${file}"`
-      file_name=`basename ${file}`
-      if [ "${operator}" = ">=" ]
+      if [ -z "${operator}" ]
       then
-        # keep if cfengine_version >= required_version
-        # exclude if cfengine_version < required_version
-        if [ "${cfengine_major}" -lt "${major}" ] || [ "${cfengine_major}" -eq "${major}" -a "${cfengine_minor}" -lt "${minor}" ]
+        include_version=0 # true for a shell
+      else
+        major=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\1/p" "${file}"`
+        minor=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\2/p" "${file}"`
+        set +e
+        if [ "${operator}" = ">=" ]
         then
-          # posix compliant syntax to exclude a file
-          excludes="${excludes} -name ${file_name} -prune -o"
+          [ "${cfengine_major}" -gt "${major}" ] || [ "${cfengine_major}" -eq "${major}" -a "${cfengine_minor}" -ge "${minor}" ]
+          include_version=$?
+        else # <
+          [ "${cfengine_major}" -lt "${major}" ] || [ "${cfengine_major}" -eq "${major}"  -a "${cfengine_minor}" -lt "${minor}" ]
+          include_version=$?
         fi
-      elif [ "${operator}" = "<" ]
+        set -e
+      fi
+
+      # @agent_requirements
+      if grep -q "${reqs_regex}" "${file}"
       then
-        # keep if cfengine_version < required_version
-        # exclude if cfengine_version >= required_version
-        if [ "${cfengine_major}" -gt "${major}" ] || [ "${cfengine_major}" -eq "${major}"  -a "${cfengine_minor}" -ge "${minor}" ]
-        then
-          # posix compliant syntax to exclude a file
-          excludes="${excludes} -name ${file_name} -prune -o"
-        fi
+        # extract capabilities from the expression
+        cap_regex='"capabilities"[ \t]*:[ \t]*\[\(.*\)\]'
+        caps=`sed -ne "s/${reqs_regex}[ \\t]*${cap_regex}[ \\t]*/\\1/p" "${file}" | tr -d '" \t' | tr ',' ' '`
+        include_reqs=0 # true for a shell
+        # for all required capability
+        for cap in ${caps}
+        do
+          count=0
+          # check if there is a matching agent capability
+          for c in ${capabilities}
+          do
+            if [ "${cap}" = "${c}" ]
+            then
+              count=1
+            fi
+          done
+          # exclue if there is no match
+          if [ ${count} -eq 0 ]
+          then
+            include_reqs=1 # false in shell
+          fi
+        done
+      else
+        # no capabilities required = accept
+        include_reqs=0 # true for a shell
+      fi
+
+      # merge and exclude if necessary
+      if [ ${include_version} -ne 0 ] || [ ${include_reqs} -ne 0 ]
```

Comment:
Be careful, this is a negative expression, and it uses shell convention for true/false.

It reads: if we exclude for version or if we exclude for requirements, then exclude ...